### PR TITLE
Exercise 3 implementation

### LINKED ISSUE #5: Step 3
Add next functions
- strdelete(unsigned int iPos, int iNumberOfSymb)
- strsubstr (unsigned int iPos, int iLen)
- find(const char* cszString)
- find(const MyString &cString)
- trim
- trimLeft
- trimRight
- substrleft(int iLength)
- substrright(int iLength)
- toUpperCase (unsigned int uiFirstsymb, unsigned int uiLastSymb)
- toLowerCase (unsigned int uiFirstsymb, unsigned int uiLastSymb)

### PR DESCRIPTION
**Pull request by Psyhich**
Closes #5

### CODE REVIEW

File: Exercise/src/my_string.cpp [hunk 1]
```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
...
+	delete[] szOldString;
+}
+
+MyStructs::CMyString MyStructs::CMyString::Substring(size_t nPosition, size_t nCharactersCount) const noexcept
+{
+	CMyString substring{nullptr}; 
+	if(nPosition >= m_nSize || nCharactersCount == 0 || 
+		m_nSize - nPosition < nCharactersCount)
+	{
+		return substring;
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:238]: You also have to print some error here.
If it is not printed in further functions call

File: Exercise/src/my_string.cpp [hunk 2]
```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
...
+	}
+	substring.TryToAllocate(nCharactersCount);
+	if(substring.m_szData == nullptr)
+	{
+		return substring;
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:244]: here also.
You need to know that something went wrong

File: Exercise/src/my_string.cpp [hunk 3]
```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
...
+	{
+		return std::nullopt;
+	}
+	
+	size_t nStringLength = GetStringLength(cszStringToFind);
+	if(nStringLength <= 1)
+	{
+		return std::nullopt;
+	}
+	nStringLength -= 1; // Excluding terminator
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:269]: What If it doesen't have \0 ?

File: Exercise/src/my_string.cpp [hunk 4]
```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
...
+	size_t nIndex = 0; // Index for iterating through this.size()
+	size_t nCurrentStringIndex = 0; // Index for iterating through cszStringToFind
+	bool bIsLookingFor = false;
+
+	for(; nIndex < m_nSize; nIndex++)
+	{
+		if(m_szData[nIndex] != cszStringToFind[nCurrentStringIndex])
+		{
+			nCurrentStringIndex = 0;
+			bIsLookingFor = false;
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: I think you can avoid this var usage in the loop
Just assign true value to this var before break in the condition

File: Exercise/src/my_string.cpp [hunk 5]
```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
...
+	return bIsLookingFor ? std::optional<size_t>(nIndex - nStringLength + 1) : std::nullopt;
+}
+
+
+MyStructs::CMyString MyStructs::CMyString::Trim(size_t nPosition, size_t nCharactersCount) noexcept
+{
+	CMyString newString{nullptr};
+	if(IsOutOfBounds(nPosition, nCharactersCount))
+	{
+		return newString;
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:302]: print something

File: Exercise/src/my_string.cpp [hunk 6]
```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
...
+void MyStructs::CMyString::ToUpperCase(size_t nStartPos, size_t nLasPos) noexcept
+{
+	if(IsOutOfBounds(nStartPos, nLasPos - nStartPos))
+	{
+		return;
+	}
+
+	for(size_t nIndex = nStartPos; nIndex < nLasPos; nIndex++)
+	{
+		m_szData[nIndex] = toupper(m_szData[nIndex]);
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:356]: Please implement this manually

File: Exercise/src/my_string.cpp [hunk 7]
```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
...
+void MyStructs::CMyString::ToLowerCase(size_t nStartPos, size_t nLasPos) noexcept
+{
+	if(IsOutOfBounds(nStartPos, nLasPos - nStartPos))
+	{
+		return;
+	}
+
+	for(size_t nIndex = nStartPos; nIndex < nLasPos; nIndex++)
+	{
+		m_szData[nIndex] = tolower(m_szData[nIndex]);
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp:369]: Please implement this manually

File: Exercise/src/my_string.cpp [hunk 8]
```diff
@@ -173,29 +173,229 @@ Exercise_1::CMyString& Exercise_1::CMyString::operator=(const Exercise_1::CMyStr
...
+		
+	}
+
+	// Manually setting terminators in case we have deleted them
+	m_szData[size() - 1] = '\0';
+	newString[newString.size() - 1] = '\0';
+
+	delete[] cszOldString;
+	return newString;
+}
```
Comment (VictorApopiy) [inline Exercise/src/my_string.cpp]: Rewrite this func according to our discussion

**Review by VictorApopiy — CHANGES_REQUESTED** [review]

File: Exercise/include/my_string.h [hunk 9]
```diff
@@ -37,6 +39,56 @@ namespace Exercise_1
...
+			Delete(nPosition, size() - nPosition);
+		}
+
+		inline void DeleteLeft(size_t nPosition) noexcept
+		{
+			Delete(0, nPosition);
+		}
+
+		CMyString Substring(size_t nPosition, size_t nCharactersCount) const noexcept;
+		inline CMyString SubstringRight(size_t nPosition) noexcept
```
Comment (VictorApopiy) [inline Exercise/include/my_string.h]: this func should be const

File: Exercise/include/my_string.h [hunk 10]
```diff
@@ -37,6 +39,56 @@ namespace Exercise_1
...
+		{
+			return Substring(nPosition, size() - nPosition);
+		}
+
+		inline CMyString SubstringLeft(size_t nPosition) noexcept
```
Comment (VictorApopiy) [inline Exercise/include/my_string.h]: this func should be const

File: Exercise/include/my_string.h [hunk 11]
```diff
@@ -37,6 +39,56 @@ namespace Exercise_1
...
+			return Substring(0, nPosition);
+		}
+
+		std::optional<size_t> Find(const char *szStringToFind) const noexcept;
+		inline std::optional<size_t> Find(const CMyString &stringToFind) const noexcept
+		{
+			return Find(stringToFind.data());
+		}
+
+		CMyString Trim(size_t nPosition, size_t nCharactersCount) noexcept;
```
Comment (VictorApopiy) [inline Exercise/include/my_string.h]: this func should be const

File: Exercise/include/my_string.h [hunk 12]
```diff
@@ -37,6 +39,56 @@ namespace Exercise_1
...
+		inline CMyString TrimRight(size_t nPosition) noexcept
```
Comment (VictorApopiy) [inline Exercise/include/my_string.h]: this func should be const

File: Exercise/include/my_string.h [hunk 13]
```diff
@@ -37,6 +39,56 @@ namespace Exercise_1
...
+		{
+			return Trim(nPosition, size() - nPosition);
+		}
+
+		inline CMyString TrimLeft(size_t nPosition) noexcept
```
Comment (VictorApopiy) [inline Exercise/include/my_string.h]: this func should be const

**Review by VictorApopiy — CHANGES_REQUESTED** [review]

File: Exercise/src/my_string.cpp (hunk 1 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:238]: Added printf to print error

File: Exercise/src/my_string.cpp (hunk 2 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:244]: Added printf to print error

File: Exercise/src/my_string.cpp (hunk 4 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Moved to break block

File: Exercise/src/my_string.cpp (hunk 5 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:302]: Added printf of error

File: Exercise/include/my_string.h (hunk 9 above)
Comment (Psyhich) [reply Exercise/include/my_string.h]: Changed to const

File: Exercise/include/my_string.h (hunk 11 above)
Comment (Psyhich) [reply Exercise/include/my_string.h]: Changed to const

File: Exercise/include/my_string.h (hunk 10 above)
Comment (Psyhich) [reply Exercise/include/my_string.h]: Changed to const

File: Exercise/include/my_string.h (hunk 12 above)
Comment (Psyhich) [reply Exercise/include/my_string.h]: Changed to const

File: Exercise/include/my_string.h (hunk 13 above)
Comment (Psyhich) [reply Exercise/include/my_string.h]: Changed to const

File: Exercise/src/my_string.cpp (hunk 3 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:269]: What should I do in such situation? I cannot check size of array or such thing without system calls

File: Exercise/src/my_string.cpp (hunk 8 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp]: Now this function trims all characters that match filter

File: Exercise/src/my_string.cpp (hunk 6 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:356]: Implemented

File: Exercise/src/my_string.cpp (hunk 7 above)
Comment (Psyhich) [reply Exercise/src/my_string.cpp:369]: Implemented

**Pull request merged**